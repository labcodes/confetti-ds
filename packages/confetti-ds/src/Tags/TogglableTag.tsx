import React, { SyntheticEvent } from "react";

import AbstractTag from "./AbstractTag";
import Icon from "../Icon";

export interface TogglableTagProps {
  /** This is the Tag's text. */
  text: string;
  /** Sets Tag's color. */
  color?: string;
  /** Sets an outline style. */
  isOutline?: boolean;
  /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled?: boolean;
  /** Disables the Tag. Won't be read by screen readers. */
  disabled?: boolean;
  /** Action to be executed when the Tag is clicked. */
  onClick?: (event?: SyntheticEvent) => any;
  /** Defines if the Tag is toggled on. */
  isOn?: boolean;
}

export default function TogglableTag({
  text,
  color,
  isOutline = false,
  ariaDisabled = false,
  disabled = false,
  onClick = () => {},
  isOn = false,
}: TogglableTagProps) {
  const selected = () => {
    let iconClass;

    if (isOn) {
      iconClass = "lab-tag--check-icon lab-tag--check-icon-on";
    } else {
      iconClass = "lab-tag--check-icon lab-tag--check-icon-off";
    }

    return <Icon type="confirm" className={iconClass} />;
  };
  return (
    <AbstractTag
      className={`lab-tag--togglable${isOn ? " lab-tag--selected" : ""}`}
      text={text}
      color={color}
      isOutline={isOutline}
      disabled={!ariaDisabled && disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      skin={isOn ? "vivid" : "pale"}
      renderPrefix={selected()}
      tabIndex={0}
    />
  );
}
