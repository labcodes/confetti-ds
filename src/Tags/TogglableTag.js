import React from "react";
import PropTypes from "prop-types";

import AbstractTag from "./AbstractTag";
import Icon from "../Icon";
import { TAG_COLORS } from "../constants";

export default class TogglableTag extends React.Component {
  static propTypes = {
    /** This is the Tag's text. */
    text: PropTypes.string.isRequired,
    /** Sets Tag's color. */
    color: PropTypes.oneOf(TAG_COLORS),
    /** Sets an outline style. */
    isOutline: PropTypes.bool,
    /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Tag. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Action to be executed when the Tag is clicked. */
    onClick: PropTypes.func,
    /** Defines if the Tag is toggled on. */
    isOn: PropTypes.bool,
  };

  static defaultProps = {
    color: "",
    isOutline: false,
    disabled: false,
    ariaDisabled: false,
    isOn: false,
    onClick: () => {},
  };

  selected = () => {
    const { isOn } = this.props;
    let iconClass;

    if (isOn) {
      iconClass = "lab-tag--check-icon lab-tag--check-icon-on";
    } else {
      iconClass = "lab-tag--check-icon lab-tag--check-icon-off";
    }

    return (
      <Icon type="check" color="black-75" size="petit" className={iconClass} />
    );
  };

  render() {
    const {
      text,
      color,
      isOutline,
      disabled,
      ariaDisabled,
      isOn,
      onClick,
    } = this.props;
    return (
      <AbstractTag
        className={`lab-tag--togglable${isOn ? " lab-tag--selected" : ""}`}
        text={text}
        color={color}
        isOutline={isOutline}
        disabled={!ariaDisabled && disabled}
        ariaDisabled={ariaDisabled}
        onClick={onClick}
        skin={isOn ? "vivid" : "pale"}
        renderPrefix={this.selected()}
      />
    );
  }
}
