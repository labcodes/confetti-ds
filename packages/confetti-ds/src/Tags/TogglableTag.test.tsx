import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import TogglableTag from "./TogglableTag";

describe("TogglableTag", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<TogglableTag text="Test render TogglableTag" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(<TogglableTag text="Test TogglableTag" />).html();
    expect(wrapper).toContain("lab-tag--togglable");
  });

  it("renders as expected with disabled as true", async () => {
    const renderedComponent = renderer
      .create(<TogglableTag text="Test nount disabled TogglableTag" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TogglableTag text="Test disabled TogglableTag" disabled />
    ).html();
    expect(wrapper).toContain("lab-tag--disabled");
  });

  it("renders as expected with outline as true", async () => {
    const renderedComponent = renderer
      .create(
        <TogglableTag text="Test render outline TogglableTag" isOutline />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TogglableTag text="Test outline TogglableTag" isOutline />
    ).html();
    expect(wrapper).toContain("lab-tag--outline");
  });

  it("renders as expected with a yellow color", async () => {
    const renderedComponent = renderer
      .create(
        <TogglableTag text="Test render yellow TogglableTag" color="yellow" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TogglableTag text="Test yellow TogglableTag" color="yellow" />
    ).html();
    expect(wrapper).toContain("lab-tag--yellow");
  });

  it("renders with a pale skin if not isOn", async () => {
    const renderedComponent = renderer
      .create(
        <TogglableTag text="Test render pale TogglableTag" isOn={false} />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TogglableTag text="Test pale TogglableTag" isOn={false} />
    ).html();
    expect(wrapper).toContain("lab-tag--pale");
  });

  it("renders with a vivid skin and a checked icon if isOn", async () => {
    const renderedComponent = renderer
      .create(<TogglableTag text="Test render isOn TogglableTag" isOn />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <TogglableTag text="Test isOn TogglableTag" isOn />
    ).html();
    expect(wrapper).toContain("lab-tag--vivid");
    expect(wrapper).toContain("lab-icon--check");
  });

  it("calls prop.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const shallowTogglableTag = shallow(
      <TogglableTag text="Test click on TogglableTag" onClick={mockOnClick} />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    shallowTogglableTag.simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it("doesn't trigger onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const mountedComponent = mount(
      <TogglableTag ariaDisabled text="Test" onClick={mockOnClick} />
    );

    expect(mockOnClick).not.toBeCalled();
    mountedComponent.find("span").at(0).simulate("click");
    expect(mockOnClick).not.toBeCalled();
  });
});
