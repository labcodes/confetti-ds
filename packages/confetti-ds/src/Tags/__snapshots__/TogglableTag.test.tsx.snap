// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TogglableTag renders as expected with a yellow color 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--yellow-pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="lab-icon lab-icon--Checkmark  lab-tag--check-icon lab-tag--check-icon-off"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.37 10.17a.5.5 0 0 0-.74.66l4 4.5c.19.22.52.23.72.02l10.5-10.5a.5.5 0 0 0-.7-.7L7.02 14.27l-3.65-4.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test render yellow TogglableTag
</span>
`;

exports[`TogglableTag renders as expected with disabled as true 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--disabled lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="lab-icon lab-icon--Checkmark  lab-tag--check-icon lab-tag--check-icon-off"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.37 10.17a.5.5 0 0 0-.74.66l4 4.5c.19.22.52.23.72.02l10.5-10.5a.5.5 0 0 0-.7-.7L7.02 14.27l-3.65-4.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test nount disabled TogglableTag
</span>
`;

exports[`TogglableTag renders as expected with outline as true 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--outline lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="lab-icon lab-icon--Checkmark  lab-tag--check-icon lab-tag--check-icon-off"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.37 10.17a.5.5 0 0 0-.74.66l4 4.5c.19.22.52.23.72.02l10.5-10.5a.5.5 0 0 0-.7-.7L7.02 14.27l-3.65-4.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test render outline TogglableTag
</span>
`;

exports[`TogglableTag renders with a pale skin if not isOn 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="lab-icon lab-icon--Checkmark  lab-tag--check-icon lab-tag--check-icon-off"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.37 10.17a.5.5 0 0 0-.74.66l4 4.5c.19.22.52.23.72.02l10.5-10.5a.5.5 0 0 0-.7-.7L7.02 14.27l-3.65-4.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test render pale TogglableTag
</span>
`;

exports[`TogglableTag renders with a vivid skin and a checked icon if isOn 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--selected lab-tag--vivid"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="lab-icon lab-icon--Checkmark  lab-tag--check-icon lab-tag--check-icon-on"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.37 10.17a.5.5 0 0 0-.74.66l4 4.5c.19.22.52.23.72.02l10.5-10.5a.5.5 0 0 0-.7-.7L7.02 14.27l-3.65-4.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test render isOn TogglableTag
</span>
`;

exports[`TogglableTag renders with base props 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="lab-icon lab-icon--Checkmark  lab-tag--check-icon lab-tag--check-icon-off"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.37 10.17a.5.5 0 0 0-.74.66l4 4.5c.19.22.52.23.72.02l10.5-10.5a.5.5 0 0 0-.7-.7L7.02 14.27l-3.65-4.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test render TogglableTag
</span>
`;
